File: python/src/taskulus/agents_management.py
"""Helpers for managing AGENTS.md Taskulus instructions."""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import List, Optional, Tuple

import click
from jinja2 import Environment

from taskulus.config_loader import ConfigurationError, load_project_configuration
from taskulus.models import ProjectConfiguration
from taskulus.project import ProjectMarkerError, get_configuration_path
from taskulus.project_management_template import (
    DEFAULT_PROJECT_MANAGEMENT_TEMPLATE,
    DEFAULT_PROJECT_MANAGEMENT_TEMPLATE_FILENAME,
)


TASKULUS_SECTION_HEADER = "## Project management with Taskulus"
TASKULUS_SECTION_LINES = [
    TASKULUS_SECTION_HEADER,
    "",
    "Use Taskulus for task management.",
    "Why: Taskulus task management is MANDATORY here; every task must live in Taskulus.",
    "When: Create/update the Taskulus task before coding; close it only after the change lands.",
    "How: See CONTRIBUTING_AGENT.md for the Taskulus workflow, hierarchy, status rules, priorities, command examples, and the sins to avoid.",
    "Warning: Editing project/ directly is a sin against The Way. Do not read or write anything in project/; work only through Taskulus.",
    "",
]
TASKULUS_SECTION_TEXT = "\n".join(TASKULUS_SECTION_LINES)
AGENTS_HEADER_LINES = ["# Agent Instructions", ""]
PROJECT_MANAGEMENT_FILENAME = "CONTRIBUTING_AGENT.md"


@dataclass(frozen=True)
class SectionMatch:
    """Represents a matched Markdown section range."""

    start: int
    end: int
    level: int


def ensure_agents_file(root: Path, force: bool) -> bool:
    """Ensure AGENTS.md exists and contains the Taskulus section.

    :param root: Repository root path.
    :type root: Path
    :param force: Whether to overwrite existing Taskulus section.
    :type force: bool
    :return: True if the file was modified.
    :rtype: bool
    :raises click.ClickException: If overwrite is required but not confirmed.
    """
    instructions_text = build_project_management_text(root)
    agents_path = root / "AGENTS.md"
    if not agents_path.exists():
        content = "\n".join(AGENTS_HEADER_LINES + TASKULUS_SECTION_LINES)
        agents_path.write_text(content, encoding="utf-8")
        _ensure_project_management_file(root, force, instructions_text)
        return True

    lines = agents_path.read_text(encoding="utf-8").splitlines()
    match = _find_taskulus_section(lines)
    if match:
        if not force:
            if not _confirm_overwrite():
                _ensure_project_management_file(root, force, instructions_text)
                _ensure_project_guard_files(root)
                return False
        updated = _replace_section(lines, match, TASKULUS_SECTION_LINES)
        agents_path.write_text(updated, encoding="utf-8")
        _ensure_project_management_file(root, force, instructions_text)
        _ensure_project_guard_files(root)
        return True

    updated_lines = _insert_taskulus_section(lines, TASKULUS_SECTION_LINES)
    agents_path.write_text(updated_lines, encoding="utf-8")
    _ensure_project_management_file(root, force, instructions_text)
    _ensure_project_guard_files(root)
    return True


def _ensure_project_management_file(
    root: Path, force: bool, instructions_text: str
) -> None:
    instructions_path = root / PROJECT_MANAGEMENT_FILENAME
    if instructions_path.exists() and not force:
        return
    instructions_path.write_text(instructions_text, encoding="utf-8")


def _ensure_project_guard_files(root: Path) -> None:
    project_dir = root / "project"
    if not project_dir.exists():
        return
    project_agents = project_dir / "AGENTS.md"
    project_agents.write_text(
        "\n".join(
            [
                "# DO NOT EDIT HERE",
                "",
                "Editing anything under project/ directly is hacking the data and is a sin against The Way.",
                "Do not read or write in this folder. Use Taskulus commands instead.",
                "",
                "See ../AGENTS.md and ../CONTRIBUTING_AGENT.md for required process.",
            ]
        )
        + "\n",
        encoding="utf-8",
    )
    do_not_edit = project_dir / "DO_NOT_EDIT"
    do_not_edit.write_text(
        "\n".join(
            [
                "DO NOT EDIT ANYTHING IN project/",
                "This folder is guarded by The Way.",
                "All changes must go through Taskulus (see ../AGENTS.md and ../CONTRIBUTING_AGENT.md).",
            ]
        )
        + "\n",
        encoding="utf-8",
    )


def build_project_management_text(root: Path) -> str:
    """Build CONTRIBUTING_AGENT.md content from configuration.

    :param root: Repository root path.
    :type root: Path
    :return: Rendered project management instructions.
    :rtype: str
    :raises click.ClickException: If configuration is missing or invalid.
    """
    configuration = _load_configuration_for_instructions(root)
    template_path = _resolve_project_management_template_path(root, configuration)
    if template_path is None:
        template_text = DEFAULT_PROJECT_MANAGEMENT_TEMPLATE
    else:
        template_text = template_path.read_text(encoding="utf-8")
    context = _build_project_management_context(configuration)
    environment = Environment(autoescape=False)
    try:
        return environment.from_string(template_text).render(context)
    except Exception as error:
        raise click.ClickException(str(error)) from error


def _load_configuration_for_instructions(root: Path) -> ProjectConfiguration:
    try:
        configuration_path = get_configuration_path(root)
        return load_project_configuration(configuration_path)
    except ProjectMarkerError as error:
        raise click.ClickException(str(error)) from error
    except ConfigurationError as error:
        raise click.ClickException(str(error)) from error


def _resolve_project_management_template_path(
    root: Path, configuration: ProjectConfiguration
) -> Optional[Path]:
    configured = configuration.project_management_template
    if configured:
        path = Path(configured)
        if not path.is_absolute():
            path = root / configured
        if not path.exists():
            raise click.ClickException(
                f"project management template not found: {path}"
            )
        return path
    conventional = root / DEFAULT_PROJECT_MANAGEMENT_TEMPLATE_FILENAME
    if conventional.exists():
        return conventional
    return None


def _build_project_management_context(
    configuration: ProjectConfiguration,
) -> dict[str, object]:
    hierarchy = configuration.hierarchy
    types = configuration.types
    workflows = _build_workflow_context(configuration.workflows)
    priorities = _build_priority_context(configuration.priorities)
    default_priority = configuration.priorities.get(configuration.default_priority)
    default_priority_name = (
        default_priority.name if default_priority else str(configuration.default_priority)
    )
    return {
        "project_key": configuration.project_key,
        "hierarchy_order": " -> ".join(hierarchy) if hierarchy else "none",
        "non_hierarchical_types": list(types),
        "parent_child_rules": _build_parent_child_rules(hierarchy, types),
        "initial_status": configuration.initial_status,
        "workflows": workflows,
        "priorities": priorities,
        "default_priority_value": configuration.default_priority,
        "default_priority_name": default_priority_name,
        "command_examples": _build_command_examples(configuration),
        "semantic_release_mapping": _build_semantic_release_mapping(types),
        "has_story": any(value.lower() == "story" for value in types),
        "gherkin_example": [
            "Feature:",
            "Scenario:",
            "Given",
            "When",
            "Then",
        ],
    }


def _build_parent_child_rules(
    hierarchy: List[str], types: List[str]
) -> List[str]:
    rules: List[str] = []
    if len(hierarchy) > 1:
        for index in range(1, len(hierarchy)):
            child = hierarchy[index]
            parent = hierarchy[index - 1]
            rules.append(f"{child} can have parent {parent}.")
    if types:
        parents = hierarchy[:-1]
        if parents:
            rules.append(
                f"{', '.join(types)} can have parent {', '.join(parents)}."
            )
        else:
            rules.append(f"{', '.join(types)} cannot have parents.")
    if len(hierarchy) <= 1 and not types:
        rules.append("No parent-child relationships are defined.")
    return rules


def _build_workflow_context(
    workflows: dict[str, dict[str, List[str]]]
) -> List[dict[str, object]]:
    context: List[dict[str, object]] = []
    for workflow_name in sorted(workflows):
        workflow = workflows[workflow_name]
        statuses = []
        for status in sorted(workflow):
            transitions = workflow[status]
            statuses.append({"name": status, "transitions": list(transitions)})
        context.append({"name": workflow_name, "statuses": statuses})
    return context


def _build_priority_context(
    priorities: dict[int, object],
) -> List[dict[str, object]]:
    context: List[dict[str, object]] = []
    for value in sorted(priorities):
        definition = priorities[value]
        context.append({"value": value, "name": definition.name})
    return context


def _build_command_examples(configuration: ProjectConfiguration) -> List[str]:
    hierarchy = configuration.hierarchy
    types = configuration.types
    priorities = sorted(configuration.priorities)
    priority_example = priorities[0] if priorities else configuration.default_priority
    workflow_name = (
        "default"
        if "default" in configuration.workflows
        else sorted(configuration.workflows)[0]
    )
    workflow = configuration.workflows[workflow_name]
    status_example = _select_status_example(
        configuration.initial_status, workflow
    )
    status_set = _collect_statuses(workflow)
    lines: List[str] = []
    if hierarchy:
        lines.append(f"tsk create \"Plan the roadmap\" --type {hierarchy[0]}")
    if len(hierarchy) > 1:
        lines.append(
            f"tsk create \"Release v1\" --type {hierarchy[1]} "
            f"--parent <{hierarchy[0]}-id>"
        )
    if len(hierarchy) > 2:
        lines.append(
            f"tsk create \"Implement feature\" --type {hierarchy[2]} "
            f"--parent <{hierarchy[1]}-id>"
        )
    if types:
        parent = hierarchy[1] if len(hierarchy) > 1 else None
        parent_fragment = f" --parent <{parent}-id>" if parent else ""
        lines.append(
            f"tsk create \"Fix crash on launch\" --type {types[0]} "
            f"--priority {priority_example}{parent_fragment}"
        )
    lines.append(
        f"tsk update <id> --status {status_example} --assignee \"you@example.com\""
    )
    if "blocked" in status_set and status_example != "blocked":
        lines.append("tsk update <id> --status blocked")
    lines.append("tsk comment <id> \"Progress note\"")
    lines.append(f"tsk list --status {configuration.initial_status}")
    lines.append("tsk close <id> --comment \"Summary of the change\"")
    return lines


def _build_semantic_release_mapping(types: List[str]) -> List[dict[str, str]]:
    mapping: List[dict[str, str]] = []
    for issue_type in types:
        lowered = issue_type.lower()
        if "bug" in lowered or "fix" in lowered:
            category = "fix"
        elif "story" in lowered or "feature" in lowered:
            category = "feat"
        elif "chore" in lowered or "maintenance" in lowered:
            category = "chore"
        else:
            category = "chore"
        mapping.append({"type": issue_type, "category": category})
    return mapping


def _collect_statuses(workflow: dict[str, List[str]]) -> set[str]:
    statuses = set(workflow.keys())
    for transitions in workflow.values():
        statuses.update(transitions)
    return statuses


def _select_status_example(initial_status: str, workflow: dict[str, List[str]]) -> str:
    if initial_status in workflow and workflow[initial_status]:
        return workflow[initial_status][0]
    for transitions in workflow.values():
        if transitions:
            return transitions[0]
    return initial_status


def _confirm_overwrite() -> bool:
    try:
        return click.confirm(
            "Taskulus section already exists in AGENTS.md. Overwrite it?",
            default=False,
        )
    except click.Abort as error:
        raise click.ClickException(
            "Taskulus section already exists in AGENTS.md. Re-run with --force to overwrite."
        ) from error


def _find_taskulus_section(lines: List[str]) -> Optional[SectionMatch]:
    for index, line in enumerate(lines):
        header = _parse_header(line)
        if not header:
            continue
        level, text = header
        if "taskulus" not in text.lower():
            continue
        end = _find_section_end(lines, index + 1, level)
        return SectionMatch(start=index, end=end, level=level)
    return None


def _find_section_end(lines: List[str], start: int, level: int) -> int:
    for index in range(start, len(lines)):
        header = _parse_header(lines[index])
        if not header:
            continue
        next_level, _ = header
        if next_level <= level:
            return index
    return len(lines)


def _parse_header(line: str) -> Optional[Tuple[int, str]]:
    match = re.match(r"^(#{1,6})\s+(.*)$", line.rstrip())
    if not match:
        return None
    level = len(match.group(1))
    text = match.group(2).strip()
    if not text:
        return None
    return level, text


def _replace_section(lines: List[str], match: SectionMatch, section_lines: List[str]) -> str:
    updated = lines[: match.start] + section_lines + lines[match.end :]
    return _join_lines(updated)


def _insert_taskulus_section(lines: List[str], section_lines: List[str]) -> str:
    insert_index = _find_insert_index(lines)
    updated = list(lines)
    if insert_index > 0 and insert_index < len(updated) and updated[insert_index].strip():
        updated.insert(insert_index, "")
        insert_index += 1
    updated[insert_index:insert_index] = section_lines
    return _join_lines(updated)


def _find_insert_index(lines: List[str]) -> int:
    for index, line in enumerate(lines):
        header = _parse_header(line)
        if header and header[0] == 1:
            insert_index = index + 1
            while insert_index < len(lines) and not lines[insert_index].strip():
                insert_index += 1
            return insert_index
    return 0


def _join_lines(lines: List[str]) -> str:
    return "\n".join(lines) + "\n"

File: rust/src/agents_management.rs
//! Helpers for managing AGENTS.md Kanbus instructions.

use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::io::{self, Write};
use std::path::Path;

#[cfg(tarpaulin)]
use crate::config::default_project_configuration;
use crate::config_loader::load_project_configuration;
use crate::error::KanbusError;
use crate::file_io::get_configuration_path;
use crate::models::ProjectConfiguration;
use crate::project_management_template::{
    DEFAULT_PROJECT_MANAGEMENT_TEMPLATE, DEFAULT_PROJECT_MANAGEMENT_TEMPLATE_FILENAME,
};
use serde::Serialize;

const KANBUS_SECTION_HEADER: &str = "## Project management with Kanbus";
const KANBUS_SECTION_LINES: [&str; 9] = [
    KANBUS_SECTION_HEADER,
    "",
    "Use Kanbus for task management.",
    "Why: Kanbus task management is MANDATORY here; every task must live in Kanbus.",
    "When: Create/update the Kanbus task before coding; close it only after the change lands.",
    "How: See CONTRIBUTING_AGENT.md for the Kanbus workflow, hierarchy, status rules, priorities, command examples, and the sins to avoid. Never inspect project/ or issue JSON directly (including with cat or jq); use Kanbus commands only.",
    "Performance: Prefer kanbusr (Rust) when available; kanbus (Python) is equivalent but slower.",
    "Warning: Editing project/ directly is a sin against The Way. Do not read or write anything in project/; work only through Kanbus.",
    "",
];
const AGENTS_HEADER_LINES: [&str; 2] = ["# Agent Instructions", ""];
const PROJECT_MANAGEMENT_FILENAME: &str = "CONTRIBUTING_AGENT.md";

#[derive(Debug, Clone)]
struct SectionMatch {
    start: usize,
    end: usize,
}

/// Ensure AGENTS.md exists and contains the Kanbus section.
///
/// # Arguments
/// * `root` - Repository root path
/// * `force` - Overwrite existing Kanbus section without prompting
///
/// # Errors
/// Returns `KanbusError::IssueOperation` if overwrite is required but not confirmed.
pub fn ensure_agents_file(root: &Path, force: bool) -> Result<(), KanbusError> {
    let instructions_text = build_project_management_text(root)?;
    let agents_path = root.join("AGENTS.md");
    if !agents_path.exists() {
        let content = build_new_agents_file();
        fs::write(&agents_path, content).map_err(|error| KanbusError::Io(error.to_string()))?;
        ensure_project_management_file(root, force, &instructions_text)?;
        ensure_project_guard_files(root)?;
        return Ok(());
    }

    let contents =
        fs::read_to_string(&agents_path).map_err(|error| KanbusError::Io(error.to_string()))?;
    let lines: Vec<String> = contents.lines().map(|line| line.to_string()).collect();
    let sections = find_kanbus_sections(&lines);
    if let Some(section) = sections.first() {
        if !force && !confirm_overwrite()? {
            ensure_project_management_file(root, force, &instructions_text)?;
            ensure_project_guard_files(root)?;
            return Ok(());
        }
        let updated = replace_sections(&lines, &sections, section, &KANBUS_SECTION_LINES);
        fs::write(&agents_path, updated).map_err(|error| KanbusError::Io(error.to_string()))?;
        ensure_project_management_file(root, force, &instructions_text)?;
        ensure_project_guard_files(root)?;
        return Ok(());
    }

    let updated = insert_kanbus_section(&lines, &KANBUS_SECTION_LINES);
    fs::write(&agents_path, updated).map_err(|error| KanbusError::Io(error.to_string()))?;
    ensure_project_management_file(root, force, &instructions_text)?;
    ensure_project_guard_files(root)?;
    Ok(())
}

/// Return the canonical Kanbus section text.
pub fn kanbus_section_text() -> String {
    let lines = KANBUS_SECTION_LINES
        .iter()
        .map(|value| value.to_string())
        .collect::<Vec<_>>();
    join_lines(&lines)
}

/// Return the Kanbus project management text derived from configuration.
///
/// # Arguments
/// * `root` - Repository root path
///
/// # Errors
/// Returns `KanbusError` if configuration lookup fails.
pub fn project_management_text(root: &Path) -> Result<String, KanbusError> {
    build_project_management_text(root)
}

fn build_project_management_text(root: &Path) -> Result<String, KanbusError> {
    let configuration_path = get_configuration_path(root)?;
    let configuration = load_project_configuration(&configuration_path)?;
    let template_path = resolve_project_management_template_path(root, &configuration)?;
    let template_text = match template_path {
        Some(path) => {
            std::fs::read_to_string(&path).map_err(|error| KanbusError::Io(error.to_string()))?
        }
        None => DEFAULT_PROJECT_MANAGEMENT_TEMPLATE.to_string(),
    };
    let context = build_project_management_context(&configuration);
    let env = minijinja::Environment::new();
    env.render_str(&template_text, context)
        .map_err(|error| KanbusError::IssueOperation(error.to_string()))
}

fn build_new_agents_file() -> String {
    let mut lines: Vec<&str> = Vec::new();
    lines.extend(AGENTS_HEADER_LINES);
    lines.extend(KANBUS_SECTION_LINES);
    join_lines(
        &lines
            .iter()
            .map(|value| value.to_string())
            .collect::<Vec<_>>(),
    )
}

#[derive(Debug, Serialize)]
struct WorkflowContext {
    name: String,
    statuses: Vec<WorkflowStatusContext>,
}

#[derive(Debug, Serialize)]
struct WorkflowStatusContext {
    name: String,
    transitions: Vec<String>,
}

#[derive(Debug, Serialize)]
struct PriorityContext {
    value: u8,
    name: String,
}

#[derive(Debug, Serialize)]
struct SemanticReleaseMapping {
    r#type: String,
    category: String,
}

#[derive(Debug, Serialize)]
struct ProjectManagementContext {
    project_key: String,
    hierarchy_order: String,
    non_hierarchical_types: Vec<String>,
    parent_child_rules: Vec<String>,
    initial_status: String,
    workflows: Vec<WorkflowContext>,
    priorities: Vec<PriorityContext>,
    default_priority_value: u8,
    default_priority_name: String,
    command_examples: Vec<String>,
    semantic_release_mapping: Vec<SemanticReleaseMapping>,
    has_story: bool,
    gherkin_example: Vec<String>,
}

fn resolve_project_management_template_path(
    root: &Path,
    configuration: &ProjectConfiguration,
) -> Result<Option<std::path::PathBuf>, KanbusError> {
    if let Some(path) = configuration.project_management_template.as_ref() {
        let resolved = if std::path::Path::new(path).is_absolute() {
            std::path::PathBuf::from(path)
        } else {
            root.join(path)
        };
        if !resolved.exists() {
            return Err(KanbusError::IssueOperation(format!(
                "project management template not found: {}",
                resolved.display()
            )));
        }
        return Ok(Some(resolved));
    }
    let conventional = root.join(DEFAULT_PROJECT_MANAGEMENT_TEMPLATE_FILENAME);
    if conventional.exists() {
        return Ok(Some(conventional));
    }
    Ok(None)
}

fn build_project_management_context(
    configuration: &ProjectConfiguration,
) -> ProjectManagementContext {
    let hierarchy = &configuration.hierarchy;
    let types = &configuration.types;
    let workflows = build_workflow_context(&configuration.workflows);
    let priorities = build_priority_context(&configuration.priorities);
    let default_priority_name = configuration
        .priorities
        .get(&configuration.default_priority)
        .map(|definition| definition.name.clone())
        .unwrap_or_else(|| configuration.default_priority.to_string());
    ProjectManagementContext {
        project_key: configuration.project_key.clone(),
        hierarchy_order: if hierarchy.is_empty() {
            "none".to_string()
        } else {
            hierarchy.join(" -> ")
        },
        non_hierarchical_types: types.clone(),
        parent_child_rules: build_parent_child_rules(hierarchy, types),
        initial_status: configuration.initial_status.clone(),
        workflows,
        priorities,
        default_priority_value: configuration.default_priority,
        default_priority_name,
        command_examples: build_command_examples(configuration),
        semantic_release_mapping: build_semantic_release_mapping(types),
        has_story: types.iter().any(|value| value.to_lowercase() == "story"),
        gherkin_example: vec![
            "Feature:".to_string(),
            "Scenario:".to_string(),
            "Given".to_string(),
            "When".to_string(),
            "Then".to_string(),
        ],
    }
}

fn build_parent_child_rules(hierarchy: &[String], types: &[String]) -> Vec<String> {
    let mut rules = Vec::new();
    if hierarchy.len() > 1 {
        for index in 1..hierarchy.len() {
            let child = &hierarchy[index];
            let parent = &hierarchy[index - 1];
            rules.push(format!("{child} can have parent {parent}."));
        }
    }
    if !types.is_empty() {
        let parents = if hierarchy.len() > 1 {
            hierarchy[..hierarchy.len() - 1].join(", ")
        } else {
            String::new()
        };
        if parents.is_empty() {
            rules.push(format!("{} cannot have parents.", types.join(", ")));
        } else {
            rules.push(format!("{} can have parent {}.", types.join(", "), parents));
        }
    }
    if hierarchy.len() <= 1 && types.is_empty() {
        rules.push("No parent-child relationships are defined.".to_string());
    }
    rules
}

fn build_workflow_context(
    workflows: &BTreeMap<String, BTreeMap<String, Vec<String>>>,
) -> Vec<WorkflowContext> {
    let mut context = Vec::new();
    for (workflow_name, workflow) in workflows {
        let mut statuses = Vec::new();
        for (status, transitions) in workflow {
            statuses.push(WorkflowStatusContext {
                name: status.clone(),
                transitions: transitions.clone(),
            });
        }
        context.push(WorkflowContext {
            name: workflow_name.clone(),
            statuses,
        });
    }
    context
}

fn build_priority_context(
    priorities: &BTreeMap<u8, crate::models::PriorityDefinition>,
) -> Vec<PriorityContext> {
    let mut context = Vec::new();
    for (value, definition) in priorities {
        context.push(PriorityContext {
            value: *value,
            name: definition.name.clone(),
        });
    }
    context
}

fn build_command_examples(configuration: &ProjectConfiguration) -> Vec<String> {
    let hierarchy = &configuration.hierarchy;
    let types = &configuration.types;
    let priority_example = configuration
        .priorities
        .keys()
        .next()
        .copied()
        .unwrap_or(configuration.default_priority);
    let workflow_name = if configuration.workflows.contains_key("default") {
        "default".to_string()
    } else {
        configuration
            .workflows
            .keys()
            .next()
            .cloned()
            .unwrap_or_else(|| "default".to_string())
    };
    let empty_workflow = BTreeMap::new();
    let workflow = configuration
        .workflows
        .get(&workflow_name)
        .unwrap_or(&empty_workflow);
    let status_example = select_status_example(&configuration.initial_status, workflow);
    let status_set = collect_statuses(workflow);
    let mut lines = Vec::new();
    if let Some(top) = hierarchy.first() {
        lines.push(format!("kanbus create \"Plan the roadmap\" --type {top}"));
    }
    if hierarchy.len() > 1 {
        lines.push(format!(
            "kanbus create \"Release v1\" --type {} --parent <{}-id>",
            hierarchy[1], hierarchy[0]
        ));
    }
    if hierarchy.len() > 2 {
        lines.push(format!(
            "kanbus create \"Implement feature\" --type {} --parent <{}-id>",
            hierarchy[2], hierarchy[1]
        ));
    }
    if let Some(issue_type) = types.first() {
        let parent = if hierarchy.len() > 1 {
            Some(&hierarchy[1])
        } else {
            None
        };
        let parent_fragment = parent
            .map(|value| format!(" --parent <{value}-id>"))
            .unwrap_or_default();
        lines.push(format!(
            "kanbus create \"Fix crash on launch\" --type {issue_type} --priority {priority_example}{parent_fragment}"
        ));
    }
    lines.push(format!(
        "kanbus update <id> --status {status_example} --assignee \"you@example.com\""
    ));
    if status_set.contains("blocked") && status_example != "blocked" {
        lines.push("kanbus update <id> --status blocked".to_string());
    }
    lines.push("kanbus comment <id> \"Progress note\"".to_string());
    lines.push(format!(
        "kanbus list --status {}",
        configuration.initial_status
    ));
    lines.push("kanbus close <id> --comment \"Summary of the change\"".to_string());
    lines
}

fn build_semantic_release_mapping(types: &[String]) -> Vec<SemanticReleaseMapping> {
    let mut mapping = Vec::new();
    for issue_type in types {
        let lowered = issue_type.to_lowercase();
        let category = if lowered.contains("bug") || lowered.contains("fix") {
            "fix"
        } else if lowered.contains("story") || lowered.contains("feature") {
            "feat"
        } else {
            "chore"
        };
        mapping.push(SemanticReleaseMapping {
            r#type: issue_type.clone(),
            category: category.to_string(),
        });
    }
    mapping
}

fn collect_statuses(workflow: &BTreeMap<String, Vec<String>>) -> BTreeSet<String> {
    let mut statuses = BTreeSet::new();
    for (status, transitions) in workflow {
        statuses.insert(status.clone());
        for transition in transitions {
            statuses.insert(transition.clone());
        }
    }
    statuses
}

fn select_status_example(initial_status: &str, workflow: &BTreeMap<String, Vec<String>>) -> String {
    if let Some(transitions) = workflow.get(initial_status) {
        if let Some(value) = transitions.first() {
            return value.clone();
        }
    }
    for transitions in workflow.values() {
        if let Some(value) = transitions.first() {
            return value.clone();
        }
    }
    initial_status.to_string()
}

fn ensure_project_management_file(
    root: &Path,
    force: bool,
    content: &str,
) -> Result<(), KanbusError> {
    let instructions_path = root.join(PROJECT_MANAGEMENT_FILENAME);
    if instructions_path.exists() && !force {
        return Ok(());
    }
    fs::write(&instructions_path, content).map_err(|error| KanbusError::Io(error.to_string()))?;
    Ok(())
}

fn ensure_project_guard_files(root: &Path) -> Result<(), KanbusError> {
    let project_dir = root.join("project");
    if !project_dir.exists() {
        return Ok(());
    }
    let project_agents = project_dir.join("AGENTS.md");
    let project_agents_content = [
        "# DO NOT EDIT HERE",
        "",
        "Editing anything under project/ directly is hacking the data and is a sin against The Way.",
        "Do not read or write in this folder. Use Kanbus commands instead.",
        "",
        "See ../AGENTS.md and ../CONTRIBUTING_AGENT.md for required process.",
    ]
    .join("\n")
        + "\n";
    fs::write(&project_agents, project_agents_content)
        .map_err(|error| KanbusError::Io(error.to_string()))?;

    let do_not_edit = project_dir.join("DO_NOT_EDIT");
    let do_not_edit_content = [
        "DO NOT EDIT ANYTHING IN project/",
        "This folder is guarded by The Way.",
        "All changes must go through Kanbus (see ../AGENTS.md and ../CONTRIBUTING_AGENT.md).",
    ]
    .join("\n")
        + "\n";
    fs::write(&do_not_edit, do_not_edit_content).map_err(|error| KanbusError::Io(error.to_string()))
}

fn confirm_overwrite() -> Result<bool, KanbusError> {
    print!("Kanbus section already exists in AGENTS.md. Overwrite it? [y/N] ");
    io::stdout()
        .flush()
        .map_err(|error| KanbusError::Io(error.to_string()))?;
    let mut input = String::new();
    let bytes = io::stdin()
        .read_line(&mut input)
        .map_err(|error| KanbusError::Io(error.to_string()))?;
    if bytes == 0 {
        return Err(KanbusError::IssueOperation(
            "Kanbus section already exists in AGENTS.md. Re-run with --force to overwrite."
                .to_string(),
        ));
    }
    let response = input.trim().to_lowercase();
    Ok(response == "y" || response == "yes")
}

fn find_kanbus_sections(lines: &[String]) -> Vec<SectionMatch> {
    let mut sections = Vec::new();
    for (index, line) in lines.iter().enumerate() {
        if let Some((level, text)) = parse_header(line) {
            if text.to_lowercase().contains("kanbus") {
                let end = find_section_end(lines, index + 1, level);
                sections.push(SectionMatch { start: index, end });
            }
        }
    }
    sections
}

fn find_section_end(lines: &[String], start: usize, level: usize) -> usize {
    for (index, line) in lines.iter().enumerate().skip(start) {
        if let Some((next_level, _)) = parse_header(line) {
            if next_level <= level {
                return index;
            }
        }
    }
    lines.len()
}

fn parse_header(line: &str) -> Option<(usize, String)> {
    let trimmed = line.trim_end_matches(&['\r', '\n'][..]);
    if !trimmed.starts_with('#') {
        return None;
    }
    let mut count = 0;
    for ch in trimmed.chars() {
        if ch == '#' {
            count += 1;
        } else {
            break;
        }
    }
    if count == 0 || count > 6 {
        return None;
    }
    let rest = &trimmed[count..];
    if !rest.starts_with(' ') && !rest.starts_with('\t') {
        return None;
    }
    let text = rest.trim();
    if text.is_empty() {
        return None;
    }
    Some((count, text.to_string()))
}

#[cfg(tarpaulin)]
pub fn cover_parse_header_cases() {
    let _ = parse_header("plain text");
    let _ = parse_header("####### too many");
    let _ = parse_header("#NoSpace");
    let _ = parse_header("# ");
    let _ = parse_header("# Header");
    let _ = parse_header("##\tTabbed");
}

#[cfg(tarpaulin)]
pub fn cover_agents_management_paths(root: &Path) {
    let mut configuration = default_project_configuration();
    configuration.project_management_template = Some("missing-template.md".to_string());
    let _ = resolve_project_management_template_path(root, &configuration);
    let absolute_template = root.join("absolute-template.md");
    let _ = fs::write(&absolute_template, "template");
    configuration.project_management_template =
        Some(absolute_template.to_string_lossy().to_string());
    let _ = resolve_project_management_template_path(root, &configuration);
    configuration.project_management_template = None;
    let _ = resolve_project_management_template_path(root, &configuration);
    let conventional = root.join(DEFAULT_PROJECT_MANAGEMENT_TEMPLATE_FILENAME);
    let _ = fs::write(&conventional, "template");
    let _ = resolve_project_management_template_path(root, &configuration);

    let _ = build_parent_child_rules(&[], &["bug".to_string()]);
    let _ = build_parent_child_rules(&[], &[]);

    let mut empty_workflows = configuration.clone();
    empty_workflows.workflows = BTreeMap::new();
    let _ = build_command_examples(&empty_workflows);
    let mut single_level = configuration.clone();
    single_level.hierarchy = vec!["task".to_string()];
    single_level.types = vec!["bug".to_string()];
    let _ = build_command_examples(&single_level);

    let mut transitions = BTreeMap::new();
    transitions.insert("open".to_string(), Vec::new());
    transitions.insert("done".to_string(), vec!["closed".to_string()]);
    let _ = select_status_example("open", &transitions);
    let _ = select_status_example("open", &BTreeMap::new());

    let _ = find_section_end(&[String::from("No headers")], 0, 1);
    let _ = parse_header("# ");
    let _ = replace_sections(
        &[String::from("# Header")],
        &[],
        &SectionMatch { start: 1, end: 1 },
        &["## Project management with Kanbus"],
    );
    let _ = find_insert_index(&[String::from("No header here")]);

    let instructions = root.join(PROJECT_MANAGEMENT_FILENAME);
    let _ = fs::write(&instructions, "content");
    let _ = ensure_project_management_file(root, false, "content");
}

fn replace_sections(
    lines: &[String],
    sections: &[SectionMatch],
    primary: &SectionMatch,
    section_lines: &[&str],
) -> String {
    let mut updated = Vec::new();
    let mut inserted = false;
    for (index, line) in lines.iter().enumerate() {
        if is_in_sections(index, sections) {
            if index == primary.start && !inserted {
                updated.extend(section_lines.iter().map(|value| value.to_string()));
                inserted = true;
            }
            continue;
        }
        updated.push(line.clone());
    }
    if !inserted {
        updated.extend(section_lines.iter().map(|value| value.to_string()));
    }
    join_lines(&updated)
}

fn is_in_sections(index: usize, sections: &[SectionMatch]) -> bool {
    sections
        .iter()
        .any(|section| index >= section.start && index < section.end)
}

fn insert_kanbus_section(lines: &[String], section_lines: &[&str]) -> String {
    let mut updated: Vec<String> = lines.to_vec();
    let mut insert_index = find_insert_index(lines);
    if insert_index > 0 && insert_index < updated.len() && !updated[insert_index].trim().is_empty()
    {
        updated.insert(insert_index, String::new());
        insert_index += 1;
    }
    let section_strings = section_lines.iter().map(|value| value.to_string());
    updated.splice(insert_index..insert_index, section_strings);
    join_lines(&updated)
}

fn find_insert_index(lines: &[String]) -> usize {
    for (index, line) in lines.iter().enumerate() {
        if let Some((level, _)) = parse_header(line) {
            if level == 1 {
                let mut insert_index = index + 1;
                while insert_index < lines.len() && lines[insert_index].trim().is_empty() {
                    insert_index += 1;
                }
                return insert_index;
            }
        }
    }
    0
}

fn join_lines(lines: &[String]) -> String {
    let mut output = lines.join("\n");
    output.push('\n');
    output
}
